servers/src/main/java/tachyon/master/block/BlockMaster.java

```diff
@@ -0,0 +1,571 @@
+      new IndexedSet.FieldIndex<MasterWorkerInfo>() {
+        @Override
+        public Object getFieldValue(MasterWorkerInfo o) {
+          return o.getAddress();
+        }
+      };
+
+  @SuppressWarnings("unchecked")
+  private final IndexedSet<MasterWorkerInfo> mWorkers =
+      new IndexedSet<MasterWorkerInfo>(mIdIndex, mAddressIndex);
+  // This state must be journaled.
+  private final AtomicLong mNextWorkerId = new AtomicLong(1);
+
+  public static String getJournalDirectory(String baseDirectory) {
+    return PathUtils.concatPath(baseDirectory, Constants.BLOCK_MASTER_SERVICE_NAME);
+  }
+
+  public BlockMaster(Journal journal, TachyonConf tachyonConf) {
+    super(journal,
+        Executors.newFixedThreadPool(2, ThreadFactoryUtils.build("block-master-%d", true)));
+    mTachyonConf = tachyonConf;
+  }
+
+  @Override
+  public TProcessor getProcessor() {
+    return new BlockMasterService.Processor<BlockMasterServiceHandler>(
+        new BlockMasterServiceHandler(this));
+  }
+
+  @Override
+  public String getProcessorName() {
+    return Constants.BLOCK_MASTER_SERVICE_NAME;
+  }
+
+  @Override
+  public void processJournalCheckpoint(JournalInputStream inputStream) throws IOException {
+    // clear state before processing checkpoint.
+    mBlocks.clear();
+
+    super.processJournalCheckpoint(inputStream);
+  }
+
+  @Override
+  public void processJournalEntry(JournalEntry entry) throws IOException {
+    if (entry instanceof BlockIdGeneratorEntry) {
+      mBlockIdGenerator.setNextContainerId(((BlockIdGeneratorEntry) entry).getNextContainerId());
+    } else if (entry instanceof WorkerIdGeneratorEntry) {
+      mNextWorkerId.set(((WorkerIdGeneratorEntry) entry).getNextWorkerId());
+    } else if (entry instanceof BlockInfoEntry) {
+      BlockInfoEntry bie = (BlockInfoEntry) entry;
+      mBlocks.put(bie.getBlockId(), new MasterBlockInfo(bie.getBlockId(), bie.getLength()));
+    } else {
+      throw new IOException("unexpected entry in journal: " + entry);
+    }
+  }
+
+  @Override
+  public void writeToJournal(JournalOutputStream outputStream) throws IOException {
+    mBlockIdGenerator.writeToJournal(outputStream);
+    outputStream.writeEntry(new WorkerIdGeneratorEntry(mNextWorkerId.get()));
+    for (MasterBlockInfo blockInfo : mBlocks.values()) {
+      outputStream.writeEntry(new BlockInfoEntry(blockInfo.getBlockId(), blockInfo.getLength()));
+    }
+  }
+
+  @Override
+  public void start(boolean isLeader) throws IOException {
+    super.start(isLeader);
+    if (isLeaderMode()) {
+      mLostWorkerDetectionService =
+          getExecutorService().submit(
+              new HeartbeatThread("Lost worker detection service",
+                  new LostWorkerDetectionHeartbeatExecutor(), mTachyonConf
+                      .getInt(Constants.MASTER_HEARTBEAT_INTERVAL_MS)));
+    }
+  }
+
+  @Override
+  public void stop() throws IOException {
+    super.stop();
+    if (isLeaderMode()) {
```

Comment:
Is this check necessary? If so, could the two checks be combined in one line (just more readable).